crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -599,3 +509,150 @@ impl TryFrom<TransactionBlockEffects> for TransactionBlock {
         })
     }
 }
+
+fn apply_scan_limited_pagination(
+    conn: &mut ScanConnection<String, TransactionBlock>,
+    page: &Page<Cursor>,
+    transactions: &Vec<StoredTransaction>,
+    tx_bounds: TxBounds,
+    scan_limit: Option<u64>,
+    checkpoint_viewed_at: u64,
+) {
+    if page.is_from_front() {
+        apply_forward_scan_limited_pagination(
+            conn,
+            page,
+            transactions,
+            tx_bounds,
+            scan_limit,
+            checkpoint_viewed_at,
+        );
+    } else {
+        apply_backward_scan_limited_pagination(
+            conn,
+            page,
+            transactions,
+            tx_bounds,
+            scan_limit,
+            checkpoint_viewed_at,
+        );
+    }
+}
+
+fn apply_forward_scan_limited_pagination(
+    conn: &mut ScanConnection<String, TransactionBlock>,
+    page: &Page<Cursor>,
+    transactions: &Vec<StoredTransaction>,
+    tx_bounds: TxBounds,
+    scan_limit: Option<u64>,
+    checkpoint_viewed_at: u64,
+) {
+    if !conn.has_previous_page {
+        // There are two special scenarios to consider when `has_previous_page` is false:
+        // 1. The cursor is scan-limited - then the corresponding element would not be in results,
+        //    which means we consequently report that there is no preceding page.
+        // 2. `scan_limit` is set, and we get an empty result set - then we report that there is no
+        //    preceding page.
+        // In both scenarios, we want to override the default behavior and set the new `startCursor`
+        // to 1 greater than the given `after` cursor.
+
+        if let Some(after) = page.after() {
```

Comment:
Requiring an `after` cursor to exist prevents us from updating the start cursor for the very first scanned page. I'm thinking in that case we should also be updating the start cursor to `tx_bounds.lo` or thereabouts (I might be off by one), but keep `has_previous_page = false`.

Initially, I was thinking that requiring `conn.has_previous_page` to be false may also cause us problems when the `after` cursor *is* found in the results, even though it is also the scan limit boundary. While that's true, I'm less worried about it because the only difference then is that we don't set the `is_scan_limited` flag, and things will still keep working in that case, but maybe there's an opportunity to simplify all this logic by taking this point and the above point into account.